### PR TITLE
[JOBS-13997] Add SdkJobsUtils and fix taskValuesUtils in proxy mode

### PR DESCRIPTION
**Pull request by vsamoilov**
## Changes
- Fix `taskValues` in proxy mode, because it is not field, but method
- Implement SdkJobsUtils according to https://docs.google.com/document/d/1yrXXA27GJOB968apgnxnMsWfj0SJJn2cpArE4zzIORU/edit#heading=h.a2at0pdkk01r

## Tests
- New tests
- Current tests

### CODE REVIEW

File: databricks-dbutils-scala/src/main/scala/com/databricks/sdk/scala/dbutils/DBUtils.scala [hunk 1]
```diff
@@ -543,19 +543,19 @@ trait DataUtils extends Serializable with WithHelpMethods {
 case class RunId private[dbutils] (id: Long)
 case class CommandContext private[dbutils] (
     /** The run ID of the root run in a workflow. */
-    rootRunId: Option[RunId],
+    rootRunId: Option[RunId] = None,
```
Comment (gaborratky-db) [inline databricks-dbutils-scala/src/main/scala/com/databricks/sdk/scala/dbutils/DBUtils.scala:546]: We should avoid introducing default values for `CommandContext` fields and initialize explicitly with these values only where necessary.

File: databricks-dbutils-scala/src/main/scala/com/databricks/sdk/scala/dbutils/SdkDBUtilsImpl.scala [hunk 2]
```diff
@@ -152,3 +162,54 @@ private class SdkSecretsUtils(client: WorkspaceClient) extends SecretUtils with
...
     }
 }
+
+private class SdkJobsUtils extends JobsUtils with NoHelp {
+  override val taskValues: TaskValuesUtils = new SdkTaskValues
+}
+
+private class SdkTaskValues extends TaskValuesUtils with NoHelp {
+  private var commandContext =
+    CommandContext()
```
Comment (gaborratky-db) [inline databricks-dbutils-scala/src/main/scala/com/databricks/sdk/scala/dbutils/SdkDBUtilsImpl.scala:172]: Do we need an instance here? It seems our integration does not rely on it. Can this be `private var commandContext: CommandContext = _`?

**Review by gaborratky-db — CHANGES_REQUESTED** [review]
We should limit our changes to introducing an implementation of `TaskValuesUtils` and not make changes to `CommandContext`.

File: databricks-dbutils-scala/src/main/scala/com/databricks/sdk/scala/dbutils/SdkDBUtilsImpl.scala [hunk 3]
```diff
@@ -20,26 +20,36 @@ private object SdkDbfsUtilsImpl {
...
 
 class SdkDBUtilsImpl(config: DatabricksConfig) extends DBUtils with NoHelp {
   private val client = new WorkspaceClient(config)
+
   def this() = this(new DatabricksConfig())
 
-  override def widgets: WidgetsUtils = SdkDbfsUtilsImpl.unsupportedField("widgets")
-  override def meta: MetaUtils = SdkDbfsUtilsImpl.unsupportedField("meta")
+  override def widgets: WidgetsUtils = SdkDBUtilsImpl.unsupportedField("widgets")
+
```
Comment (gaborratky-db) [inline databricks-dbutils-scala/src/main/scala/com/databricks/sdk/scala/dbutils/SdkDBUtilsImpl.scala:33]: nit: the newlines are unnecessary diff. can we keep the code blocks as intended by previous author?

File: databricks-dbutils-scala/src/main/scala/com/databricks/sdk/scala/dbutils/SdkDBUtilsImpl.scala [hunk 4]
```diff
@@ -152,3 +162,54 @@ private class SdkSecretsUtils(client: WorkspaceClient) extends SecretUtils with
...
+  override def getJson(taskKey: String, key: String): Seq[String] = Seq.empty
+
+  override def getContext(): CommandContext = {
+    commandContext
+  }
+
+  override def setContext(context: CommandContext): Unit = {
+    commandContext = context
+    println(context)
+    println(commandContext)
```
Comment (gaborratky-db) [inline databricks-dbutils-scala/src/main/scala/com/databricks/sdk/scala/dbutils/SdkDBUtilsImpl.scala:213]: leftover debug statements?

File: databricks-dbutils-scala/src/test/scala/com/databricks/sdk/scala/dbutils/integration/SdkTaskValuesIntegrationTest.scala [hunk 5]
```diff
@@ -0,0 +1,65 @@
+package com.databricks.sdk.scala.dbutils.integration
+
+import com.databricks.sdk.scala.dbutils.CommandContext
+import org.scalatest.matchers.should.Matchers.convertToAnyShouldWrapper
+import org.scalatest.matchers.should.Matchers.be
+
+class SdkTaskValuesIntegrationTest extends IntegrationTestBase {
+  "When outside of DBR, TaskValuesUtils.getContext" should "return default context initially" taggedAs Integration in {
+    if (isInDbr) {
+      cancel("This test must only be run inside DBR")
```
Comment (gaborratky-db) [inline databricks-dbutils-scala/src/test/scala/com/databricks/sdk/scala/dbutils/integration/SdkTaskValuesIntegrationTest.scala:10]:
```suggestion
      cancel("This test must only be run outside DBR")
```

File: databricks-dbutils-scala/src/test/scala/com/databricks/sdk/scala/dbutils/integration/SdkTaskValuesIntegrationTest.scala [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    dbutils.jobs.taskValues.getContext should be(CommandContext())
+  }
+
+  "When outside of DBR, TaskValuesUtils.getContext" should "return context, which was set" taggedAs Integration in {
+    if (isInDbr) {
+      cancel("This test must only be run inside DBR")
```
Comment (gaborratky-db) [inline databricks-dbutils-scala/src/test/scala/com/databricks/sdk/scala/dbutils/integration/SdkTaskValuesIntegrationTest.scala:18]:
```suggestion
      cancel("This test must only be run outside DBR")
```

File: databricks-dbutils-scala/src/test/scala/com/databricks/sdk/scala/dbutils/integration/SdkTaskValuesIntegrationTest.scala [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    val getE = intercept[NotImplementedError] {
+      dbutils.jobs.taskValues.get("test", "test", None, None)
+    }
+
+    getE.getMessage should be("get is not supported in Scala.")
+  }
+
+  "When outside of DBR, TaskValuesUtils.setJson" should "throw error" taggedAs Integration in {
+    if (isInDbr) {
+      cancel("This test must only be run inside DBR")
```
Comment (gaborratky-db) [inline databricks-dbutils-scala/src/test/scala/com/databricks/sdk/scala/dbutils/integration/SdkTaskValuesIntegrationTest.scala:48]:
```suggestion
      cancel("This test must only be run outside DBR")
```

File: databricks-dbutils-scala/src/test/scala/com/databricks/sdk/scala/dbutils/integration/SdkTaskValuesIntegrationTest.scala [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    val e = intercept[NotImplementedError] {
+      dbutils.jobs.taskValues.setJson("test", "test")
+    }
+
+    e.getMessage should be("setJson is not supported for local development in Scala.")
+  }
+
+  "When outside of DBR, TaskValuesUtils.getJson" should "return empty list" taggedAs Integration in {
+    if (isInDbr) {
+      cancel("This test must only be run inside DBR")
```
Comment (gaborratky-db) [inline databricks-dbutils-scala/src/test/scala/com/databricks/sdk/scala/dbutils/integration/SdkTaskValuesIntegrationTest.scala:60]:
```suggestion
      cancel("This test must only be run outside DBR")
```

File: databricks-dbutils-scala/src/main/scala/com/databricks/sdk/scala/dbutils/SdkDBUtilsImpl.scala (hunk 2 above)
Comment (vsamoilov) [reply databricks-dbutils-scala/src/main/scala/com/databricks/sdk/scala/dbutils/SdkDBUtilsImpl.scala:172]: I check that initially I return empty context

File: databricks-dbutils-scala/src/main/scala/com/databricks/sdk/scala/dbutils/SdkDBUtilsImpl.scala [hunk 9]
```diff
@@ -9,7 +9,7 @@ import java.io.ByteArrayInputStream
 import java.nio.charset.StandardCharsets
 import scala.collection.JavaConverters._
 
-private object SdkDbfsUtilsImpl {
+private object SdkDBUtilsImpl {
```
Comment (mgyucht) [inline databricks-dbutils-scala/src/main/scala/com/databricks/sdk/scala/dbutils/SdkDBUtilsImpl.scala:12]: Thanks!

File: databricks-dbutils-scala/src/main/scala/com/databricks/sdk/scala/dbutils/SdkDBUtilsImpl.scala [hunk 10]
```diff
@@ -152,3 +154,52 @@ private class SdkSecretsUtils(client: WorkspaceClient) extends SecretUtils with
...
+   * @param debugValue
+   *   the value to return when called outside of a job context (must not be None)
+   * @return
+   *   the task value (if it exists) when called inside of a job context
+   */
+  override def get(taskKey: String, key: String, default: Option[Any], debugValue: Option[Any]): Any =
+    throw new NotImplementedError("get is not supported in Scala.")
+
+  override def setJson(key: String, value: String): Unit =
+    throw new NotImplementedError("setJson is not supported for local development in Scala.")
```
Comment (gaborratky-db) [inline databricks-dbutils-scala/src/main/scala/com/databricks/sdk/scala/dbutils/SdkDBUtilsImpl.scala:194]: `setJson` should not throw an error, otherwise, code that runs on DBR would throw an error during local development. Instead, we want parity, so we simply swallow it. See the [two-pager](https://docs.google.com/document/d/1yrXXA27GJOB968apgnxnMsWfj0SJJn2cpArE4zzIORU/edit#bookmark=id.j15664rxdtvz) for requirements.

**Review by gaborratky-db — CHANGES_REQUESTED** [review]
`setJson` should not throw, otherwise the SDK cannot be used for local development when the customer relies on `setJson` on top of DBR.

I will defer final review to @mgyucht.

File: databricks-dbutils-scala/src/main/scala/com/databricks/sdk/scala/dbutils/SdkDBUtilsImpl.scala [hunk 11]
```diff
@@ -152,3 +154,52 @@ private class SdkSecretsUtils(client: WorkspaceClient) extends SecretUtils with
...
+   *   the value to return when called inside of a job context if the task value does not exist (must not be None)
+   * @param debugValue
+   *   the value to return when called outside of a job context (must not be None)
+   * @return
+   *   the task value (if it exists) when called inside of a job context
+   */
+  override def get(taskKey: String, key: String, default: Option[Any], debugValue: Option[Any]): Any =
+    throw new NotImplementedError("get is not supported in Scala.")
+
+  override def setJson(key: String, value: String): Unit =
```
Comment (mgyucht) [inline databricks-dbutils-scala/src/main/scala/com/databricks/sdk/scala/dbutils/SdkDBUtilsImpl.scala:193]: According to @gaborratky-db's doc, this should not throw an error but should be a no-op.

File: databricks-dbutils-scala/src/main/scala/com/databricks/sdk/scala/dbutils/SdkDBUtilsImpl.scala [hunk 12]
```diff
@@ -152,3 +154,52 @@ private class SdkSecretsUtils(client: WorkspaceClient) extends SecretUtils with
...
+
+  override def getJson(taskKey: String, key: String): Seq[String] = Seq.empty
+
+  override def getContext(): CommandContext = {
+    commandContext
+  }
+
+  override def setContext(context: CommandContext): Unit = {
+    commandContext = context
+  }
```
Comment (mgyucht) [inline databricks-dbutils-scala/src/main/scala/com/databricks/sdk/scala/dbutils/SdkDBUtilsImpl.scala:204]: Why do we implement these in the SDK? CommandContext is purely a DBR concept so it should be provided by DBR?

File: databricks-dbutils-scala/src/test/scala/com/databricks/sdk/scala/dbutils/integration/SdkTaskValuesIntegrationTest.scala [hunk 13]
```diff
@@ -0,0 +1,76 @@
+package com.databricks.sdk.scala.dbutils.integration
+
+import com.databricks.sdk.scala.dbutils.CommandContext
+import org.scalatest.matchers.should.Matchers.convertToAnyShouldWrapper
+import org.scalatest.matchers.should.Matchers.be
+
+class SdkTaskValuesIntegrationTest extends IntegrationTestBase {
```
Comment (mgyucht) [inline databricks-dbutils-scala/src/test/scala/com/databricks/sdk/scala/dbutils/integration/SdkTaskValuesIntegrationTest.scala:7]: This doesn't need to be an integration test since none of these tests run in DBR or interact with the Databricks REST API. Instead, let's move it up one directory into the `com.databricks.sdk.scala.dbutils` package and extend AnyFlatSpec like ProxyDbfsTest.

**Review by mgyucht — COMMENTED** [review]
Couple comments, thanks for the contribution!

File: databricks-dbutils-scala/src/main/scala/com/databricks/sdk/scala/dbutils/SdkDBUtilsImpl.scala (hunk 10 above)
Comment (vsamoilov) [reply databricks-dbutils-scala/src/main/scala/com/databricks/sdk/scala/dbutils/SdkDBUtilsImpl.scala:194]: Oh, yeah, totally misread this

File: databricks-dbutils-scala/src/main/scala/com/databricks/sdk/scala/dbutils/ProxyDBUtilsImpl.scala [hunk 14]
```diff
@@ -198,21 +200,16 @@ class ProxyDBUtilsImpl private[dbutils] (baseObj: AnyRef) extends DBUtils {
   override val library: LibraryUtils = getProxyInstance[LibraryUtils](baseObj.getField("library"))
   override val credentials: DatabricksCredentialUtils =
     getProxyInstance[DatabricksCredentialUtils](baseObj.getField("credentials"))
-  override val jobs: JobsUtils = new ProxyJobsUtils(baseObj.getField("jobs"))
+  override val jobs: JobsUtils = getProxyInstance[JobsUtils](
```
[the remark this answers is not in the document]
Comment (vsamoilov) [reply databricks-dbutils-scala/src/main/scala/com/databricks/sdk/scala/dbutils/ProxyDBUtilsImpl.scala:203]: So `taskValues` is method and not a field, so the previous implementation wasn't working with following error: `NoSuchFieldException: taskValues`

File: databricks-dbutils-scala/src/main/scala/com/databricks/sdk/scala/dbutils/SdkDBUtilsImpl.scala (hunk 11 above)
Comment (vsamoilov) [reply databricks-dbutils-scala/src/main/scala/com/databricks/sdk/scala/dbutils/SdkDBUtilsImpl.scala:193]: Yes, fixed

File: databricks-dbutils-scala/src/main/scala/com/databricks/sdk/scala/dbutils/SdkDBUtilsImpl.scala (hunk 12 above)
Comment (vsamoilov) [reply databricks-dbutils-scala/src/main/scala/com/databricks/sdk/scala/dbutils/SdkDBUtilsImpl.scala:204]: I don't know, but it is a part of public interface now, should I throw error?

Comment (vsamoilov) [reply databricks-dbutils-scala/src/main/scala/com/databricks/sdk/scala/dbutils/SdkDBUtilsImpl.scala]: Throw error, that it is not supported outside of DBR, is it OK @gaborratky-db?

File: databricks-dbutils-scala/src/main/scala/com/databricks/sdk/scala/dbutils/ProxyDBUtilsImpl.scala (hunk 14 above)
[the remark this answers is not in the document]
Comment (mgyucht) [reply databricks-dbutils-scala/src/main/scala/com/databricks/sdk/scala/dbutils/ProxyDBUtilsImpl.scala:203]: Interesting. It is defined as a method in `ProxyJobUtils`, so I don't exactly know why that didn't work, but I don't see any issue with this.

**Review by mgyucht — APPROVED** [review]
